src/rgw/rgw_aio.cc

```diff
@@ -0,0 +1,96 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include <type_traits>
+#include "include/rados/librados.hpp"
+
+#include "librados/librados_asio.h"
+
+#include "rgw_aio.h"
+
+namespace rgw {
+namespace {
+void cb(librados::completion_t, void* arg);
+}
+struct state {
+  Aio* aio;
+  librados::AioCompletion* c;
+
+  state(Aio* aio, AioResult& r)
+    : aio(aio),
+      c(librados::Rados::aio_create_completion(&r, nullptr, &cb)) {}
+};
+
+namespace {
+void cb(librados::completion_t, void* arg) {
+  static_assert(sizeof(AioResult::user_data) >= sizeof(state));
+  static_assert(std::is_trivially_destructible_v<state>);
+  auto& r = *(static_cast<AioResult*>(arg));
+  auto s = reinterpret_cast<state*>(&r.user_data);
+  r.result = s->c->get_return_value();
+  s->c->release();
+  s->aio->put(r);
+}
+
+template<typename T,
+	 bool read = std::is_same_v<std::decay_t<T>,
+				    librados::ObjectReadOperation>,
+	 typename = std::enable_if_t<std::is_base_of_v<librados::ObjectOperation,
+						       std::decay_t<T>> &&
+				     !std::is_lvalue_reference_v<T> &&
+				     !std::is_const_v<T>>>
+Aio::OpFunc aio_abstract(T&& op,
+			 optional_yield y) {
+  return [op = std::move(op),
+	  y = std::move(y)](Aio* aio, AioResult& r) mutable {
+	   if (y) {
+	     auto& context = y.get_io_context();
+	     auto& yield = y.get_yield_context();
+	     auto& ioctx = r.obj.get_ref().ioctx;
+	     auto& oid = r.obj.get_ref().obj.oid;
+	     boost::system::error_code ec;
+	     if constexpr (read) {
+	       r.data = librados::async_operate(context, ioctx, oid, &op, 0,
+						yield[ec]);
+	     } else {
+	       librados::async_operate(context, ioctx, oid, &op, 0, yield[ec]);
+	     }
```

Comment:
this looks close, but using `yield` directly as the completion token isn't quite what we want here - that would suspend the coroutine until this single io completes. instead, we want to submit several ios in parallel, and only suspend when `YieldingAioThrottle::get()`/`wait()`/`drain()` call `async_wait()` - ie where the BlockingAioThrottle would wait on its condition variable

to accomplish this, `YieldingAioThrottle::submit()` was passing a custom `YieldingAioThrottle::Handler` to `librados::async_operate()`, using `bind_executor()` to make sure it runs on the same (strand) executor as the coroutine/yield_context - this lets it avoid the locking that's otherwise needed for `BlockingAioThrottle::put()`

maybe its better to leave this piece out for now, and i'll rework https://github.com/ceph/ceph/pull/26173 later on top? up to you

Reply:
Yeah, I'll just take it out for now.